File: app/src/main/java/com/example/flightstats/data/StatsCalculator.java
package com.example.flightstats.data;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Calculates hero stats from a list of Flight objects.
 */
public class StatsCalculator {

    public int totalFlights;
    public int uniqueCountries;
    public int uniqueAirports;
    public int uniqueRoutes;
    public double totalDistanceKm;

    // Airport → country mapping is looked up from the WebView's airport data.
    // For the native hero stats we track airports and routes only.
    public static StatsCalculator compute(List<Flight> flights, AirportRepository airports) {
        StatsCalculator s = new StatsCalculator();
        s.totalFlights = flights.size();

        Set<String> airportSet = new HashSet<>();
        Set<String> routeSet = new HashSet<>();
        Set<String> countrySet = new HashSet<>();

        for (Flight f : flights) {
            airportSet.add(f.origin);
            airportSet.add(f.destination);
            // Canonical route (sorted so AMS-OSL == OSL-AMS)
            String[] pair = new String[]{f.origin, f.destination};
            java.util.Arrays.sort(pair);
            routeSet.add(pair[0] + "-" + pair[1]);
            s.totalDistanceKm += f.distance;

            if (airports != null) {
                String originCountry = airports.getCountry(f.origin);
                String destCountry = airports.getCountry(f.destination);
                if (originCountry != null) countrySet.add(originCountry);
                if (destCountry != null) countrySet.add(destCountry);
            }
        }

        s.uniqueAirports = airportSet.size();
        s.uniqueRoutes = routeSet.size();
        s.uniqueCountries = countrySet.size();
        return s;
    }
}

File: app/src/main/java/com/example/flightstats/data/StatsCalculator.kt
package com.example.flightstats.data

import java.util.Arrays

data class StatsCalculator(
    val totalFlights: Int = 0,
    val uniqueCountries: Int = 0,
    val uniqueAirports: Int = 0,
    val uniqueRoutes: Int = 0,
    val totalDistanceKm: Double = 0.0
) {
    companion object {
        fun compute(flights: List<Flight>, airports: AirportRepository?): StatsCalculator {
            val airportSet = mutableSetOf<String>()
            val routeSet = mutableSetOf<String>()
            val countrySet = mutableSetOf<String>()
            var distance = 0.0

            for (f in flights) {
                val o = f.origin.orEmpty()
                val d = f.destination.orEmpty()
                airportSet.add(o)
                airportSet.add(d)

                // Canonical route (sorted so AMS-OSL == OSL-AMS)
                val pair = arrayOf(o, d)
                pair.sort()
                routeSet.add("${pair[0]}-${pair[1]}")
                distance += f.distance

                if (airports != null) {
                    airports.getCountry(o)?.let { countrySet.add(it) }
                    airports.getCountry(d)?.let { countrySet.add(it) }
                }
            }

            return StatsCalculator(
                totalFlights = flights.size,
                uniqueAirports = airportSet.size,
                uniqueRoutes = routeSet.size,
                uniqueCountries = countrySet.size,
                totalDistanceKm = distance
            )
        }
    }
}
